Add support for AD8460

MarielTinaco opened the pull request:
This adds device tree bindings compatible strings, driver support and ZedBoard device trees for AD8460 High Voltage, High Current Arbitrary Waveform Generator DAC.

For more information:
[Arbitrary Waveform Generator with Integrated 14-Bit High Speed DAC](https://www.analog.com/media/en/technical-documentation/data-sheets/ad8460.pdf
)

This has been tested on the ZedBoard which controls the EVAL-AD8460SDZ evaluation board

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+
+	state->sync_clk = sync_clk;
+
+	vrefio = devm_regulator_get_optional(&spi->dev, "vrefio");
```
Is this property in the bindings?

nunojsa replied:
Yeah, this needs to be on the bindings...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+			     IIO_SHARED_BY_ALL, ad8460_pattern_memory_get,
+			     ad8460_pattern_memory_set),
+	AD8460_CHAN_EXT_INFO("span", 0, IIO_SHARED_BY_ALL, ad8460_span_get,
+			     NULL),
```
all these needs to be justified and need to be documented in the ABI file.... So the span is not the same as the scale?

MarielTinaco replied:
You're right, span does seem to mean the same thing is scale. On that note, I've included scale in the available channel attributes and removed span from external info.

```
        case IIO_CHAN_INFO_SCALE:
		*val = 80 * (2000 / state->rset_ohms) * (state->vref_mv / 1000);

		return IIO_VAL_INT;
	default:
		return -EINVAL;

```

nunojsa replied:
Yeah, that's the right thing to do :+1: . And yes, if you look to some other drivers, we do have same thing (span == scale)

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+			     IIO_SHARED_BY_ALL, ad8460_pattern_memory_get,
+			     ad8460_pattern_memory_set),
+	AD8460_CHAN_EXT_INFO("pattern_depth", AD8460_PATTERN_DEPTH,
+			     IIO_SHARED_BY_ALL, ad8460_pattern_memory_get,
```
What is the usecase for `pattern_depth` and `pattern_memory`?

nunojsa replied:
I'm mostly thinking on what's the usecase for this? Why would we want to have any "hand written" pattern outputting the dac? Typically, for things like patterns that you can set in the device registers, I would say it's a debug feature. That means, you could support it in degugfs. Look [here](https://github.com/analogdevicesinc/linux/blob/main/drivers/iio/adc/ad4630.c#L1411) for an example (this is a bit more complicated as there's buffering involved).

That said, you can try to push this upstream as a normal sysfs ABI (with proper ABI and justifications) but I have strong suspicious for it to be accepted.

Another thing you can do is to ask the apps responsible for this part what would be the usecase for it or if this is mostly a debug feature.

MarielTinaco replied:
It does seem like a debug feature. It's an alternative way to push out hand-written pattern through the DAC using some allocated SPI registers. But the actual way of pushing data out of the DAC is through a parallel interface of 14 pins. For which the FPGA board used is critical. I think this is the case where the buffering would be critical.

This definitely looks like the right way to implement this debug feature. I'll look into this example

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+
+		reg = FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, 0);
+		return regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x02),
+					  AD8460_PATTERN_DEPTH_MSK, reg);
```
don't we need a lock in here?

MarielTinaco replied:
Added guard lock inside the write raw function
```
	guard(mutex)(&state->lock);

	switch (mask) {
	case IIO_CHAN_INFO_ENABLE:
		return ad8460_chan_enable(state, val);
	case IIO_CHAN_INFO_RAW:
		ret = ad8460_set_hvdac_byte(state, 0, val);
		if (ret)
			return ret;

		reg = FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, 0);
		return regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x02),
					  AD8460_PATTERN_DEPTH_MSK, reg);
	default:
		return -EINVAL;
	}
```

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+	switch (mask) {
+	case IIO_CHAN_INFO_ENABLE:
+		*val =  state->enable;
+
```
what's the enable vs the powerdown state?

MarielTinaco replied:
In this case, enable directly allows/prohibits DAC output, while powerdown sets the device to low power mode by activating the sleep registers. There would still be DAC output even if powerdown is set to HIGH.

nunojsa replied:
Then I think you're mixing things :). This `enable` actually looks like it should be your [powerdown](https://elixir.bootlin.com/linux/latest/source/Documentation/ABI/testing/sysfs-bus-iio#L730). It dows not exactly fits what is stated in the docs but it is the attr that should disable the output.

And low power mode should likely be implemented using the linux power monitoring subsystem.

MarielTinaco replied:
Could the linux power monitoring subsystem be used to take care of the SLEEP registers? I am aware of the PMBUS protocol but I was told it is derived from I2C interface, whereas this device is only SPI. Are there examples of drivers making use of the linux power monitoring subsystem. Thanks, it would help a lot

MarielTinaco replied:
> Then I think you're mixing things :). This `enable` actually looks like it should be your [powerdown](https://elixir.bootlin.com/linux/latest/source/Documentation/ABI/testing/sysfs-bus-iio#L730). It dos not exactly fits what is stated in the docs but it is the attr that should disable the output.
> 
> And low power mode should likely be implemented using the linux power monitoring subsystem.

scanning the includes, there are pm_ modules that are available such as pm.h, pm_runtime.h, etc. I also saw them being used in other drivers. Are these what you were referring as the power management subsystem?

nunojsa replied:
yes.... that's the stuff that runs for example when you hibernate your laptop or put it in suspend mode. There's also things like runtime pm where you can put your in device in low power mode after some time. Couple of examples:

(system wide PM)[https://elixir.bootlin.com/linux/latest/source/drivers/iio/temperature/ltc2983.c#L1742)
[runtime PM](https://elixir.bootlin.com/linux/latest/source/drivers/iio/temperature/mlx90635.c#L1016)

If you're unsure what to use, go with the system wide one...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+
+	span = 80 * (2000 / state->rset_ohms) * (state->vref_mv / 1000);
+
+	return sysfs_emit(buf, "%ld\n", span);
```
yeah, this looks like scale

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+	if (ret)
+		return ret;
+
+	mutex_lock(&state->lock);
```
you can use the cleanup.h magic...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+	if (ret)
+		return ret;
+
+	return regmap_write(state->regmap, AD8460_HVDAC_DATA_WORD_HIGH(index),
```
This should definitely be locked... Imagine you write the lower word and get preemped and then someone else enters this path?

MarielTinaco replied:
For this, I just added the cleanup.h magic. Would this suffice?

```
static int ad8460_set_hvdac_byte(struct ad8460_state *state,
				 int index,
				 int val)
{
	int ret;

	guard(mutex)(&state->lock);

	ret = regmap_write(state->regmap, AD8460_HVDAC_DATA_WORD_LOW(index),
			   (val & 0xFF));
	if (ret)
		return ret;

	return regmap_write(state->regmap, AD8460_HVDAC_DATA_WORD_HIGH(index),
			    ((val >> 8) & 0xFF));
}

```

nunojsa replied:
yes...

MarielTinaco commented on Documentation/devicetree/bindings/iio/dac/adi,ad8460.yaml:
```diff
@@ -0,0 +1,98 @@
+    items:
+      - const: sync_clk
+
+  adi,rset-ohms:
```
the rset-ohms property corresponds to the value of the external resistor to adjust the full scale range of the HV output, as shown in the excerpt below:

![image](https://github.com/analogdevicesinc/linux/assets/104679546/0a989470-4233-4c44-8725-08ef8bcc9fa3)

It was referred to as RSET in the datasheet. I probably have to rename it into something clearer to the user. Do you have a naming suggestion?

nunojsa replied:
Hmm if it's like that in the datasheet, then I would leave it like that

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+	ret = device_property_read_u32(&spi->dev, "adi,rset-ohms",
+				       &state->rset_ohms);
+	if (ret)
+		state->rset_ohms = 2000;
```
I've used the RSET value mounted on the EVAL-AD8460SDZ board as the default value for RSET. It does have a nominal range so it couldn't be just any value

nunojsa replied:
That was my point... In your code you pretty much allow for any value to be given in the bindings. You need to make sure what's the constrains for that resistor (if any) and make sure to error out for an invalid value

MarielTinaco replied:
I ended up implementing it this way off of one example I've found in the device "da7280". Would this be considered good?

```	
        ret = device_property_read_u32(&spi->dev, "adi,rset-ohms", &temp);
	if (!ret && (temp >= 2000 && temp <= 20000))
		state->rset_ohms = temp;
	else
		return dev_err_probe(&spi->dev, -EINVAL,
					     "Invalid value for rset\n");
```

MarielTinaco replied:
Noticed that this also throws the probe error even if the error is caused by the ret variable. With that, I found another example.

```

	ret = device_property_read_u32(&spi->dev, "adi,rset-ohms", &temp);
	if (ret)
		return ret;
	if (temp < 2000 || temp > 20000)
		return dev_err_probe(&spi->dev, -EINVAL,
					     "Invalid value for rset\n");

```

MarielTinaco replied:
Found probably an even better example which I based on the driver you've mentioned in a separate comment. I think I'll go with this

```
	ret = device_property_read_u32(&spi->dev, "adi,rset-ohms", &temp);
	if (!ret) {
		if (temp < 2000 || temp > 20000)
			return dev_err_probe(&spi->dev, -EINVAL,
					"Invalid value for rset: %u\n", temp);
		state->rset_ohms = temp;
	}
```

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+}
+
+static struct iio_chan_spec_ext_info ad8460_ext_info[] = {
+	IIO_ENUM("input_mode", IIO_SEPARATE, &ad8460_input_mode_enum),
```
It's a property of the AD8460 Basically, there are two input modes, one is to write the sequence of values through SPI (APG mode) and the other is to write it through parallel interface (AWG mode)

should this be moved to the device tree?

nunojsa replied:
So it seems to me that APG mode only makes sense for the pattern thingy? If so, and if we move to debugfs, then this should the enable attribute for test_pattern.

MarielTinaco commented on Documentation/devicetree/bindings/iio/dac/adi,ad8460.yaml:
```diff
@@ -0,0 +1,98 @@
+    oneOf:
+      - items:
+          - enum: [50000, 100000, 300000, 500000, 1000000]
+          - enum: [50000, 100000, 300000, 500000, 1000000]
```
The reason for this is because the parameter is a 2-element array, but both elements have the same typical values

```
                adi,ilim-microamp = <100000 100000>;
```
Is there a better implementation for this

nunojsa replied:
Hmm, I see...

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,668 @@
+		ret = ad8460_arm(state, AD8460_OVERCURRENT_SRC, 0x1, tmp[1]);
+		if (ret)
+			return dev_err_probe(&spi->dev, -EINVAL,
+					     "overcurrent src not valid: %d",
```
Applied the same input value error handling here
```
	/* Arm the device by default */
	ret = device_property_read_u32_array(&spi->dev, "adi,ilim-microamp",
					     tmp, ARRAY_SIZE(tmp));
	if (!ret) {
		if (tmp[0] < 0 || tmp[0] > AD8460_MAX_OVERCURRENT_MICROAMP)
			return dev_err_probe(&spi->dev, -EINVAL,
					     "overcurrent snk not valid: %d",
					     tmp[0]);

		ret = ad8460_arm(state, AD8460_OVERCURRENT_SNK, 0x1, tmp[0]);
		if (ret)
			return ret;

		if (tmp[1] < 0 || tmp[1] > AD8460_MAX_OVERCURRENT_MICROAMP)
			return dev_err_probe(&spi->dev, -EINVAL,
					     "overcurrent src not valid: %d",
					     tmp[1]);

		ret = ad8460_arm(state, AD8460_OVERCURRENT_SRC, 0x1, tmp[1]);
		if (ret)
			return ret;
	}
```

MarielTinaco commented:
v2

1. Moved some attributes to debugfs
2. Applied cleanup magic
3. Provided missing includes

nunojsa commented:
@MarielTinaco, sorry for the delay. I did replied to some your questions... Make sure to fix the build errors CI is reporting and push another version of the PULL.

Hint for the future:
If you push directly to ADI git tree and name your branch as `staging/...`, basic CI will run on your dev branch so you can catch build errors like this before opening (or force push) the PULL . If you do not have access to push into ADI tree, we can have a look.

MarielTinaco commented:
Hi @nunojsa unfortunately I do not have access to push into the ADI tree. But it will surely come in handy in the future. Thanks!

MarielTinaco commented:
V3

- Fixed CI build errors

MarielTinaco commented:
V4

1. Added buffer capability

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+	ret = devm_add_action_or_reset(&spi->dev, ad8460_clk_disable, sync_clk);
+	if (ret)
+		return ret;
+
```
No need for any of the above... Note `devm_clk_get_enabled()` will already prepare_enable your clock

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+		return ret;
+
+	ret = regmap_read(state->regmap, AD8460_CTRL_REG(0x0E), &flag);
+	iio_device_release_direct_mode(indio_dev);
```
Hmm will a read() do any arm on the device if buffering?

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+	if (ret)
+		return ret;
+
+	ret = regmap_read(state->regmap, AD8460_CTRL_REG(0x02), &mode);
```
again... is a read() that bad?

MarielTinaco replied:
right, I don't think this should interfere with the buffer capture. I should just remove the device claim/release, is that right?

```
	struct iio_dev *indio_dev = arg;
	struct ad8460_state *state = iio_priv(indio_dev);
	u32 reg;
	int ret;

	ret = regmap_read(state->regmap, AD8460_CTRL_REG(0x02), &reg);
	if (ret)
		return ret;

	*val = FIELD_GET(AD8460_PATTERN_DEPTH_MSK, reg);

	return 0;
```

nunojsa replied:
> right, I don't think this should interfere with the buffer capture.

With the above in mind, I think so, yes...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+}
+
+static const struct iio_chan_spec ad8460_channels[] = {
+	AD8460_VOLTAGE_CHAN(0)
```
your macro calls it voltage but then you use `IIO_ALTVOLTAGE()`. Consistency please :)

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+	if (ret)
+		return ret;
+
+	return 0;
```
ditto

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+			   int *val,
+			   int *val2,
+			   long mask)
+{
```
can we only get samples while buffering? Or can we just read a single sample (using IIO_CHAN_INFO_RAW)?

nunojsa replied:
Ups, forget this. This is a DAC... And from a quick look on DS, it looks like we don't have any register to control the word to send out (just using that pattern memory stuff)

> Pattern depth and pattern memory are configurable settings for the Arbitrary Pattern Generator (APG) mode. Pattern memory is an array like property that is found on the SPI registers.

But taking your words, I'm now thinking if it would make sense for use to support IIO_CHAN_INFO_RAW where pattern depth is pretty much 1 and we just output a DC voltage. Would that work or does it make sense to you?

MarielTinaco replied:
Yes, it does make sense to support IIO_CHAN_INFO_RAW this way. 

I'll include this in the next revision

nunojsa replied:
yeah, this device has some funny stuff :)

nunojsa reviewed (commented):
Main thing we should see now is that limits stuff. I'm not sure but maybe somehow using the hwmon subsystem would make sense. Like that we could actually control those limits over userspace. But for that, I think we would need to be capable to also read the "thing" we're putting limits on...

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+		ret = ad8460_arm(state, AD8460_OVERTEMPERATURE, 0x1, temp);
+		if (ret)
+			return ret;
+	}
```
Yup, it does have hw monitoring capabilities. there are alarm flags upon detection of overvoltage, overcurrent and overtemp and shutdown flags, the limits can also be set throught SPI

I'll be looking into the hwmon bridge for this

nunojsa replied:
Never used it myself but here it is the main module:

https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/iio_hwmon.c

But from a quick look into the code, it does not look like it handles alarms and limits. So not really sure what is the best way here. I think you have two options...

a) Do like you have right now and see what the IIO maintainer has to say;
b) Have a second hwmon driver that implements the hwmon related bit and uses the IIO inkernal interface (there's on [driver](https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/ntc_thermistor.c) doing that already).

IMO, b) is the way to go but it's also more work. So what I would do in your place would be to drop the hwmon capabilities for now and just include the IIO part when sending the patches upstream. However, in the cover letter, I would mention all of this and ask for guidance on the way to go. The point is, I think you split this work now. First just go with the IIO driver and then if the way to go is to have a second hwmon, do that one after the IIO one is accepted. OTOH, if the IIO maintainer says, "yeah, include that hwmon interface in the IIO driver", then you can do that in the next iterations of your patchseries.

MarielTinaco replied:
Oh, sorry about that 😅. Maybe what I'm trying to say is, if there's an SPI register bit that monitors an alarm flag, is it a standard practice to poll this flag, then trigger an interrupt request if a certain value of the flag is detected.

MarielTinaco replied:
> Never used it myself but here it is the main module:
> 
> https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/iio_hwmon.c
> 
> But from a quick look into the code, it does not look like it handles alarms and limits. So not really sure what is the best way here. I think you have two options:
> 
> a) Do like you have right now and see what the IIO maintainer has to say; b) Have a second hwmon driver that implements the hwmon related bits and uses the IIO inkernal interface (there's one [driver](https://elixir.bootlin.com/linux/latest/source/drivers/hwmon/ntc_thermistor.c) doing that already).
> 
> IMO, b) is the way to go but it's also more work. So what I would do in your place would be to drop the hwmon capabilities for now and just include the IIO part when sending the patches upstream. However, in the cover letter, I would mention all of this and ask for guidance on the way to go. The point is, I think you should split this work now. First just go with the IIO driver and then if the way to go is to have a second hwmon driver, do that one after the IIO one is accepted. OTOH, if the IIO maintainer says, "yeah, include that hwmon interface in the IIO driver", then you can do that in the next iterations of your patchseries.

Now I'm thinking again, I'd like to push through the integration of the fault/alarm monitoring into the hwmon interface, which I'll ask of the IIO maintainers how to properly implement. But don't you think that the setting of alarm limits and turning on fault protection of the device in the probe function, would be nice to have, to protect the device by default. So this is like option a) but doesn't contradict with including the hwmon interface if the IIO maintainers would find necessary. What do you think?

nunojsa replied:
>  But don't you think that the setting of alarm limits and turning on fault protection of the device in the probe function, would be nice to have, to protect the device by default. 

Why not starting with the default limits? As for enabling fault protection, that could be done from the hwmon driver during probe. 

That said, I'm not sure you can get away by having the hwmon driver as a simple platform device depending on the IIO driver (using the iio inkern interface for reading the channel values). Or if you'll need a full MFD (Multi function device) driver with a common API for the IIO and hwmon drivers...

MarielTinaco replied:
> Why not starting with the default limits? As for enabling fault protection, that could be done from the hwmon driver during probe.

The default min/max limits of all fault protection (overcurrent, overvoltage, and overtemp) are set to 0 in the SPI reg, so if I enable the fault protection, before setting the fault protection limits, the device will automatically shut down. That's why fault protection is disabled by default.

If there's a way in the hwmon driver to expose the min/max limits as configurable attributes in the userspace, as well as the fault protection enable, that would probably be the way to go.

But for now, I'll just go with the defaults and remove the fault monitoring related stuff from probe function

> That said, I'm not sure you can get away by having the hwmon driver as a simple platform device depending on the IIO driver (using the iio inkern interface for reading the channel values). Or if you'll need a full MFD (Multi function device) driver with a common API for the IIO and hwmon drivers...

Found the MFD core functions, in <linux/mfd/core.h>. Are there particular examples I could use?

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+#include <linux/dmaengine.h>
+#include <linux/iio/buffer.h>
+#include <linux/iio/buffer-dma.h>
+#include <linux/iio/buffer-dmaengine.h>
```
added IIO_BUFFER and IIO_BUFFER_DMAENGINE to Kconfig entry

```
config AD8460
	tristate "Analog Devices AD8460 DAC driver"
	depends on SPI
	select IIO_BUFFER
	select IIO_BUFFER_DMAENGINE
```

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+		state->rset_ohms = temp;
+	}
+
+	/* Soft reset */
```
do you mean accessing the reset pin through GPIO controls? Although it does have a soft reset pin, it behaves the same way as the soft reset bit in the SPI register. That's why I went on with the SPI-configurable option. 

Should I support GPIO controls as well for the reset pin?

nunojsa replied:
Yes, typically that's how we do it:

https://elixir.bootlin.com/linux/latest/source/drivers/iio/dac/ltc2688.c#L867

You should name the gpio as in the datasheet... Note that the idea is that we ask the gpio in the asserted state (GPIOD_OUT_HIGH) so that if we have the gpio, we get it we the device already in reset. Then we just have to deassert it.

nunojsa replied:
Note that you need to see in datasheet for proper sleep times (don't copy it from other devices)

MarielTinaco replied:
I see. I found the correct timing of Manual RESET on the specifications. The Typical minimum duration is 10 ns

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+	if (ret)
+		return ret;
+
+	if (IS_ENABLED(CONFIG_DEBUG_FS))
```
I see. I changed return type of debugfs_init from void to int. then called it before device register.

```
	ret = ad8460_debugfs_init(indio_dev);
	if (ret)
		return ret;

	return devm_iio_device_register(&spi->dev, indio_dev);
```

```
static int ad8460_debugfs_init(struct iio_dev *indio_dev)
{
	struct dentry *d = iio_get_debugfs_dentry(indio_dev);

	if (!IS_ENABLED(CONFIG_DEBUG_FS))
		return 0;
	debugfs_create_file_unsafe("apg_mode_enable", 0600, d,
				   indio_dev, &ad8460_apg_mode_en_fops);
	debugfs_create_file_unsafe("apg_pattern_depth", 0600, d,
				   indio_dev, &ad8460_apg_pattern_depth_fops);
	debugfs_create_file_unsafe("shutdown_flag", 0600, d,
				   indio_dev, &ad8460_shutdown_flag_fops);
	debugfs_create_file("apg_pattern_memory", 0644, d,
			    indio_dev, &ad8460_apg_pattern_memory_fops);

	return 0;
}
```

nunojsa replied:
Look at your code :wink: ... Why do you have a return to only ever `return 0`? Turn it void again... we don't care about error handling when creating debugfs interfaces. Even if those fail, they are debug stuff so we should not stop our device from probing because of them...

commodo commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+	if (ret)
+		return ret;
+
+	ret = ad8460_enable_apg_mode(state, val);
```
`iio_device_release_direct_mode(indio_dev)`  here?
otherwise we leave it claimed on error;

MarielTinaco replied:
i take it that I should bump the `iio_device_release_direct_mode(indio_dev)` up before return like here?

```
	ret = ad8460_enable_apg_mode(state, val);
	iio_device_release_direct_mode(indio_dev);

	return ret;
```

commodo replied:
yep

nunojsa replied:
Well, when upstreaming use the new [iio_device_claim_direct_scoped()](https://elixir.bootlin.com/linux/v6.9-rc4/C/ident/iio_device_claim_direct_scoped). Then, no need to care about releasing it :)

commodo commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+	ret = regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x02),
+				 AD8460_PATTERN_DEPTH_MSK,
+				 FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, val));
+	if (ret)
```
`iio_device_release_direct_mode(indio_dev)`  here?
otherwise we leave it claimed on error;

MarielTinaco replied:
same as the other comment

```
	ret = regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x02),
				 AD8460_PATTERN_DEPTH_MSK,
				 FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, val));
	iio_device_release_direct_mode(indio_dev);

	return ret;
```

commodo replied:
yep

commodo commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+static struct iio_chan_spec_ext_info ad8460_ext_info[] = {
+	AD8460_CHAN_EXT_INFO("powerdown", 0, IIO_SEPARATE,
+			     ad8460_read_powerdown, ad8460_write_powerdown),
+	{},
```
general nitpick (upstream will say): null terminators, don't need a trailing comma

commodo commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,724 @@
+
+		return IIO_VAL_INT;
+	case IIO_CHAN_INFO_SCALE:
+		*val = 80 * (2000 / state->rset_ohms) * (state->vref_mv / 1000);
```
i wonder if it's better to rewrite this formula to prefer multiplications first and divisions later;
integer division tends to lose some precision;
then there can be a comment about the original formula, or reference it to the datasheet

MarielTinaco replied:
I refactored it this way
```
	case IIO_CHAN_INFO_SCALE:
		/* vCONVENTIONAL = 80V * (DAC_CODE / 2**14) - 40V
		 * vMAX = 80V * (2**14 / 2**14) - 40V
		 * vMIN = 80V * (0 / 2**14) - 40V
		 * vADJ = vCONVENTIONAL * (2000 / rSET) * (vREF / 1.2)
		 * vSPAN = vADJ_MAX - vADJ_MIN
		 * See datasheet page 49, section FULL-SCALE REDUCTION
		 */
		num = 80 * 2000 * state->vref_mv;
		denom = state->rset_ohms * 1200;
		*val = DIV_ROUND_CLOSEST_ULL(num, denom);

		return IIO_VAL_INT;
```

MarielTinaco commented on Documentation/devicetree/bindings/iio/dac/adi,ad8460.yaml:
```diff
@@ -0,0 +1,90 @@
+
+  clock-names:
+    items:
+      - const: sync_clk
```
if I remove the usage of clock-names? how does this affect the way I call sync_clk from the driver?

in the driver code, this is the current implementation
```
	sync_clk = devm_clk_get_enabled(&spi->dev, "sync_clk");
	if (IS_ERR(sync_clk))
		return dev_err_probe(&spi->dev, PTR_ERR(sync_clk),
				     "Failed to get sync clk\n");

```

MarielTinaco commented:
V5
	- remove clock-names
	- refactor debugfs init
	- handling of GPIO soft reset
	- add raw value as channel attribute
	- refactor scale calculation

	- held off pm_runtime functions due to boot errors
	- retained setting of alarm limits and arming device

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,753 @@
+		.sign = 'u',					\
+		.realbits = 14,					\
+		.storagebits = 16,				\
+		.shift = 0,					\
```
no need for shift = 0

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,753 @@
+}
+DEFINE_DEBUGFS_ATTRIBUTE(ad8460_apg_pattern_depth_fops,
+			 ad8460_show_apg_pattern_depth,
+			 ad8460_set_apg_pattern_depth, "%llu\n");
```
Do we still want the debugfs interface now we have it in write raw? The only justification I see would be to have more words.

MarielTinaco replied:
I think the pattern_depth can remain on debugfs interface. If setting two values at a time, (pattern index and word) is available in write raw, I think pattern_memory can also be removed from debugfs. I'll remove the apg_mode_enable from debugfs now, since it is handled on write raw and pre-enable of buffer.

MarielTinaco replied:
another option is to be able selection of test patterns instead such as square, ramp, default etc. this can cover both pattern_depth and pattern_memory setting. what do you think?

nunojsa replied:
Hmm not sure what you mean about the last question? Could you clarify it a bit?

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,753 @@
+	ret = clk_prepare_enable(sync_clk);
+	if (ret)
+		return dev_err_probe(&spi->dev, ret,
+				     "failed to enable trigger clk\n");
```
drop the above `clk_prepare_enable()`

MarielTinaco replied:
I see. Dropped the probe error as well

```
	sync_clk = devm_clk_get_enabled(&spi->dev, "sync_clk");
	if (IS_ERR(sync_clk))
		return dev_err_probe(&spi->dev, PTR_ERR(sync_clk),
				     "Failed to get sync clk\n");

	state->sync_clk = sync_clk;
```

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,753 @@
+		if (ret)
+			return ret;
+	}
+	usleep_range(100, 150);
```
I would use `fsleep()`

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,753 @@
+
+		reg = FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, 0);
+		return regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x02),
+					  AD8460_PATTERN_DEPTH_MSK, reg);
```
I see, it seems like a should put locking on the higher level functions or those functions that wrap the other function calls.

MarielTinaco commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,753 @@
+	int ret;
+
+	switch (mask) {
+	case IIO_CHAN_INFO_RAW:
```
Will I be able to use iio_device_claim_direct_scoped() right now? Or should I just have it during upstream and use iio_device_claim_direct_mode/iio_device_release_direct_mode for now?

MarielTinaco replied:
For now, this is what write raw looks like using the IIO device claim/release

```
	struct ad8460_state *state = iio_priv(indio_dev);
	unsigned int reg;
	int ret;

	switch (mask) {
	case IIO_CHAN_INFO_RAW:
		ret = iio_device_claim_direct_mode(indio_dev);
		if (ret)
			return ret;

		guard(mutex)(&state->lock);

		ret = ad8460_enable_apg_mode(state, 1);
		if (ret)
			return ret;

		ret = ad8460_set_hvdac_byte(state, 0, val);
		if (ret)
			return ret;

		reg = FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, 0);
		ret = regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x02),
					 AD8460_PATTERN_DEPTH_MSK, reg);
		iio_device_release_direct_mode(indio_dev);

		return ret;
	default:
		return -EINVAL;
	}
```

nunojsa replied:
> Will I be able to use iio_device_claim_direct_scoped() right now? is it supported in this repo yet? Or did you mean that I must have it during upstream and use iio_device_claim_direct_mode/iio_device_release_direct_mode for now?

I think it's only upstream for now...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,610 @@
+		if (ret)
+			return ret;
+	}
+	fsleep(100);
```
nit: you can move the above code to its own function... ad8460_reset()

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,610 @@
+static struct iio_chan_spec_ext_info ad8460_ext_info[] = {
+	AD8460_CHAN_EXT_INFO("powerdown", 0, IIO_SEPARATE,
+			     ad8460_read_powerdown, ad8460_write_powerdown),
+	{}
```
note you may be asked to support [powerdown-mode](https://elixir.bootlin.com/linux/latest/source/Documentation/ABI/testing/sysfs-bus-iio#L697)

MarielTinaco replied:
I see, although there is no SPI register that explicitly points to powerdown-modes, how would I support this?

MarielTinaco replied:
I see. Although, unlike other devices where the powerdown modes are explicitly stated as part of the SPI register, in this device, I don't think there is an exact equivalent, that comes close to the selections defined in the ABI for powerdown-modes.

MarielTinaco replied:
Now that you mention it, it does seem to have a powerdown state when HV sleep is initiated. Device OUT is terminated to 27kohms

![image](https://github.com/analogdevicesinc/linux/assets/104679546/411112f6-4604-4a81-ae8c-ea1935062e87)

"27kohms_to_gnd" doesn't fall under the available selections in the ABI, 

![image](https://github.com/analogdevicesinc/linux/assets/104679546/cdd9c247-335d-46d4-94ad-5bfc60b5c593)

are we allowed to create our own entry?

nunojsa replied:
Yes, we're allowed to update the ABI file... You can propose a new entry and support it in the drier and see how it goes :)

MarielTinaco replied:
gotcha! But since there is only possible powerdown-mode, do I just create an empty setter/getter for it? like here

```
static const char * const ad8460_powerdown_modes[] = {
	"27kohm_to_gnd",
};

static int ad8460_get_powerdown_mode(struct iio_dev *indio_dev,
				     const struct iio_chan_spec *chan)
{
	return 0;
}

static int ad8460_set_powerdown_mode(struct iio_dev *indio_dev,
				     const struct iio_chan_spec *chan,
				     unsigned int type)
{
	return 0;
}

static const struct iio_enum ad8460_powerdown_mode_enum = {
	.items = ad8460_powerdown_modes,
	.num_items = ARRAY_SIZE(ad8460_powerdown_modes),
	.get = ad8460_get_powerdown_mode,
	.set = ad8460_set_powerdown_mode,
};
```

I found one such example. [stm32-dac](https://github.com/analogdevicesinc/linux/blob/fbd2b84ac9126f3b1a4cb8f02ff93fb40311b2e0/drivers/iio/dac/stm32-dac.c#L187-L203)

nunojsa replied:
Yeah, but I don't think you're mode is `27kohm_to_gnd`. In the datasheet pic you pasted nothing states any ground connection. It says the output goes into high-z (with an impedance of around 27kohm). And that's effectively the `three_state` mode already supported.

MarielTinaco replied:
Oh, I see. You're right

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,610 @@
+	unsigned int num, denom;
+	int data, ret;
+
+	guard(mutex)(&state->lock);
```
Why not moving this into the IIO_CHAN_INFO_RAW case?

MarielTinaco replied:
If clean up magic is moved directly inside the "raw" case, as seen below

```
	switch (mask) {
	case IIO_CHAN_INFO_RAW:
		guard(mutex)(&state->lock);

		ret = ad8460_get_hvdac_byte(state, 0, &data);
		if (ret)
			return ret;

		*val = data;
		return IIO_VAL_INT;
```

An error is thrown because there is no declaration before it. So what I did is placed the guard(mutex).. inside the get_hvdac_byte

nunojsa replied:
Use:

```
scoped_guard(mutex, &state->lock) {
    ret = ad8460_get_hvdac_byte(state, 0, &data);
    if (ret)
        return ret;
}
...
```

and it should work...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,610 @@
+{
+	struct ad8460_state *state = iio_priv(indio_dev);
+
+	return ad8460_enable_apg_mode(state, 1);
```
don't you need to guard() the above two calls?

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,610 @@
+		if (ret)
+			return ret;
+
+		guard(mutex)(&state->lock);
```
nit: not a big deal but from what I understand we don't really need the lock for ad8460_enable_apg_mode() - that is protected by the iio internal lock - so we could move it before ad8460_set_hvdac_byte()

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,610 @@
+	int sdn_flag;
+	int ret;
+
+	guard(mutex)(&state->lock);
```
nit: move this after `kstrtobool()`

MarielTinaco replied:
```
	int ret;

	ret = kstrtobool(buf, &pwr_down);
	if (ret)
		return ret;

	guard(mutex)(&state->lock);

	ret = regmap_update_bits(state->regmap, AD8460_CTRL_REG(0x01),
				 AD8460_HVDAC_SLEEP_MSK,
				 FIELD_PREP(AD8460_HVDAC_SLEEP_MSK, pwr_down));
```

nunojsa reviewed (commented):
This already looks good enough to send a first version upstream. So you can take my last review and send it out :). Couple of things to note:

* You need to improve your commit message on the patch adding support for the driver. Don't mention `via axi`. You need a small description of the device (like you have in the bindings patch);
* Don't forget to mention the hwmon capabilities and how can we better support them (if still have the intuit of adding them)
* And be prepared for some (maybe lots) changes to be asked. This is a very funny part so I expect the upstream review to be equally "funny" :)

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,656 @@
+	}
+	fsleep(100);
+
+	return ret;
```
I think `ret` can be used without any initialization in case gpio is used.

MarielTinaco replied:
Refactored the reset function this way. To remove the dependency to `ret`
```
static int ad8460_reset(const struct ad8460_state *state)
{
	struct device *dev = &state->spi->dev;
	struct gpio_desc *gpio;

	gpio = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_LOW);
	if (IS_ERR(gpio))
		return dev_err_probe(dev, PTR_ERR(gpio),
				     "Failed to get reset gpio");
	if (gpio) {
		fsleep(100);
		gpiod_set_value_cansleep(gpio, 1);
	} else {
		return regmap_write(state->regmap, AD8460_CTRL_REG(0x03), 1);
	}
	fsleep(100);

	return 0;
}
```

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,656 @@
+	if (ret)
+		return ret;
+
+	*val = (FIELD_GET(AD8460_DATA_BYTE_HIGH_MSK, high) << 8) | low;
```
double check it but I don't think the parenthesis...

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,656 @@
+	char data[16];
+	int ret, val;
+
+	guard(mutex)(&state->lock);
```
use scoped_guard() and use it in the only place needed...  `ad8460_set_hvdac_byte()`

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,656 @@
+{
+	struct ad8460_state *state = iio_priv(indio_dev);
+
+	guard(mutex)(&state->lock);
```
AFAICT, you don't need the mutex lock here. ad8460_enable_apg_mode() only runs against your write_raw() function and the call is protected with iio_device_claim_direct_mode() which used the same internal IIO mutex as the postdisable() and preenable() callbacks. So in this case, the internal IIO lock is fine to guarantee correctness.

nunojsa commented on drivers/iio/dac/ad8460.c:
```diff
@@ -0,0 +1,656 @@
+
+		ret = ad8460_enable_apg_mode(state, 1);
+		if (ret)
+			return ret;
```
You should use the scoped version of iio_device_claim_direct_mode() when upstreaming. But note that in this code you have a deadlock as you're returning without iio_device_release_direct_mode()

MarielTinaco replied:
I ended up with this refactor

```
	int ret;

	switch (mask) {
	case IIO_CHAN_INFO_RAW:
		ret = iio_device_claim_direct_mode(indio_dev);
		if (ret)
			return ret;

		ret = ad8460_enable_apg_mode(state, 1);
		iio_device_release_direct_mode(indio_dev);
		if (ret)
			return ret;

		scoped_guard(mutex, &state->lock) {
			ret = ad8460_set_hvdac_byte(state, 0, val);
			if (ret)
				return ret;

			reg = FIELD_PREP(AD8460_PATTERN_DEPTH_MSK, 0);
			ret = regmap_update_bits(state->regmap,
						 AD8460_CTRL_REG(0x02),
						 AD8460_PATTERN_DEPTH_MSK, reg);
		}

		return ret;
	default:
		return -EINVAL;
	}
```

nunojsa reviewed (commented):
Small notes from me... go ahead and send it upstream (I'll also be on vacations for two weeks so no need to wait.)

MarielTinaco commented:
V7

Applied changes from upstream

MarielTinaco commented:
Change logs

V7

1. Added and enabled other voltage regulators: HVEE, HVCC, AVDD_3P3V, DVDD_3P3V, VREF_5V, VCC_5V
2. Added overtemp, under/overvoltage and under/overcurrent fault monitoring as IIO Events
3. Added programmable current channel
4. Added optional temperature read, using IIO consumer channel
5. Added entry for Shutdown and Reset GPIO's
6. Enabled Arbitrary Pattern Generator (APG) mode by setting channel attribute "raw[0-15]". Can also be controlled by "toggle_en" and "symbol" attribute

nunojsa commented:
@MarielTinaco,

Could you update the PR by cherry-picking directly the commits from upstream? It then makes it obvious that this is upstream. Also, don't we have dt bindings for this?
